### PR TITLE
doc: initial draft of Assembly coding guidelines

### PR DESCRIPTION
**Pull request by shiqingg**
This patch creates initial draft of Assembly coding guidelines.

Signed-off-by: Shiqing Gao <shiqing.gao@intel.com>

### CODE REVIEW

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+
+       asm_showcase_1:
+       asm_showcase_2:
+               movl    $0x1, %eax
+
+
+ASM-GN-02: Name reserved for use by the assembler shall not be used for any other purpose
+=========================================================================================
+
+Name defined by developers shall not be the same as assembler directive,
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]: Better would be:
```
Names defined by developers shall not be the same as an assembler directive,
instruction prefix, instruction mnemonic, or register name.
```

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       asm_showcase_1:
+       asm_showcase_2:
+               movl    $0x1, %eax
+
+
+ASM-GN-02: Name reserved for use by the assembler shall not be used for any other purpose
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]:
```suggestion
ASM-GN-02: Names reserved for use by the assembler shall not be used for any other purpose
```

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+       .section .data
+       showcase_data:
+               /* 0x1234 means <something> */
+               .long   0x1234 * 0x1234 >> 2
+
+
+ASM-GN-06: .end directive statement shall be the last statement in an Assembly file
+===================================================================================
+
+If .end directive is used in an Assembly file, it shall be the last statement.
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]: Can you write an assembly file without a ``.end`` directive?   The guideline says the last statement must be an ``.end`` directive, so saying "If" here doesn't seem right.

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+       execution_32:
+               /* do something in 32-bit mode */
+               ljmp    $0x0008, $asm_showcase_data
+
+
+ASM-GN-10: Assembler directives shall be used with restrictions
+===============================================================
+
+Usage of the assembler directive refers to GNU assembler 'as' user manual. Only
+the following assembler directives could be used:
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]: change ``could`` to ``may``

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       .CODE64
+       LOCK AND        %RCX, (%RDX)
+
+
+ASM-NC-02: Name defined by developers shall use lower case letters
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]:
```suggestion
ASM-NC-02: Names defined by developers shall use lower case letters
```

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+==================================================================
+
+Name defined by developers shall use lower case letters with the following
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]:
```suggestion
Names defined by developers shall use lower case letters with the following
```

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,771 @@
...
+   Non-compliant example::
+
+       asm_showcase:
+               movl    $0x1, %eax
+       
+       asm_showcase:
+               movl    $0x2, %eax
+
+
+ASM-NC-04: Name defined by developers shall be less than 31 characters
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]:
```suggestion
ASM-NC-04: Names defined by developers shall be less than 31 characters
```

File: doc/developer-guides/asm_coding_guidelines.rst (hunk 3 above)
Comment (shiqingg) [reply doc/developer-guides/asm_coding_guidelines.rst]: `.end` directive is not a must-have for assembly file.
If it is used in an assembly file, it shall be the last statement.
If it is not used, this rule does not apply.

I updated the detailed description in V2.

**Comment by shiqingg** [comment]
@dbkinder Thanks for your review David!
New version has been pushed. Some comments are inline.
Could you please help review it again?
Thanks!

File: doc/developer-guides/asm_coding_guidelines.rst [hunk 8]
```diff
@@ -0,0 +1,772 @@
...
+   Non-compliant example::
+
+       // This is a comment
+       movl    $0x1, %eax
+
+
+ASM-CS-06: Tabs shall be 8 characters wide
+==========================================
+
+A tab character shall be considered 8-character wide when limiting the line
```
Comment (dbkinder) [inline doc/developer-guides/asm_coding_guidelines.rst]: I don't see mention of a line length limit (80 or 120 perhaps)?

Comment (shiqingg) [reply doc/developer-guides/asm_coding_guidelines.rst]: > I don't see mention of a line length limit (80 or 120 perhaps)?

@dbkinder Thanks for pointing out!
Just added this rule in new version. 
Could you please help review it again?
Thanks!

**Pull request merged**